Update user guides and docs regarding tasks using Java toolchains (#22985)
gradle/gradle: subprojects/docs/src/docs/userguide/jvm/building_java_projects.adoc

=== Targeting a specific Java version

By default, Gradle will compile Java code to the language level of the JVM running Gradle.

> If you need to target a specific version of Java when compiling, Gradle provides multiple options.

Reviewer: With the list that follows, should this end with `:`?

1. Using <<toolchains.adoc#toolchains,Java toolchains>> is a preferred way to target a language version. +

> A toolchains uniformly handles compilation, execution and Javadoc generation, and it can be configured on the project level.

Reviewer 2: ```suggestion
A toolchain uniformly handles compilation, execution, and Javadoc generation and can be configured on the project level.
```

2. Using link:{groovyDslPath}/org.gradle.api.tasks.compile.CompileOptions.html#org.gradle.api.tasks.compile.CompileOptions:release[`release`] property is possible starting from Java 10. +
Selecting a Java release makes sure that compilation is done with the configured language level and against the JDK APIs from that Java version.
3. Using `sourceCompatibility` and `targetCompatibility` properties. +

> Although not generally advised, these options were historically used to configure Java version during compilation.

Reviewer 2: ```suggestion
Although not generally advised, these options were historically used to configure the Java version during compilation.
```

[[sec:compiling_with_toolchain]]
==== Using toolchains

When Java code is compiled using a specific toolchain, the actual compilation is carried out by a compiler of the specified Java version.
The compiler provides access to the language features and JDK APIs for the requested Java language version.

> In the simplest case, the toolchain can be configured for a project using `java` extension.

Reviewer 2: ```suggestion
In the simplest case, the toolchain can be configured for a project using the `java` extension.
```

This way, not only compilation benefits from it, but also other tasks such as `test` and `javadoc` using the toolchain consistently.

====
include::sample[dir="snippets/java/toolchain-basic/kotlin",files="build.gradle.kts[tags=toolchain]"]
include::sample[dir="snippets/java/toolchain-basic/groovy",files="build.gradle[tags=toolchain]"]
====

You can learn more about this in the <<toolchains.adoc#toolchains,Java toolchains>> guide.

[[sec:compiling_with_release]]
==== Using Java release version

Setting a link:{groovyDslPath}/org.gradle.api.tasks.compile.CompileOptions.html#org.gradle.api.tasks.compile.CompileOptions:release[release] flag makes sure that the specified language level is used regardless of which compiler actually performs the compilation.
Although, the compiler has to support the requested release version.

> It is possible to specify an earlier release version, while compiling with a more recent <<#sec:compiling_with_toolchain,toolchain>>.

Reviewer 2: ```suggestion
It is possible to specify an earlier release version while compiling with a more recent <<#sec:compiling_with_toolchain,toolchain>>.
```

Gradle supports using the release flag from Java 10.
It can be configured on the compilation task as follows.

.Setting Java release flag
====
include::sample[dir="snippets/java/basic/kotlin",files="build.gradle.kts[tags=java-release-flag]"]
include::sample[dir="snippets/java/basic/groovy",files="build.gradle[tags=java-release-flag]"]
====

The release flag provides guarantees similar to toolchains.
It validates that the Java sources are not using language features introduced in later Java versions, and also that the code does not access APIs from more recent JDKs.
The bytecode produced by the compiler also corresponds to the release Java version, meaning that the compiled code cannot be executed on older JVMs.

Technically, the `release` option of the Java compiler was introduced in Java 9.

> However, using this option with Gradle is only possible starting with Java 10, due to a https://bugs.openjdk.java.net/browse/JDK-8139607[bug in Java 9].

Author: Yep, this is how I've seen the external links formatted in the docs. I also checked that it works in the local build of the docs.
